src/Adapter/Image/Uploader/ProductImageUploader.php

```diff
@@ -0,0 +1,167 @@
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Image\Uploader;
+
+use ErrorException;
+use Image;
+use ImageType;
+use PrestaShop\PrestaShop\Adapter\Image\ImageGenerator;
+use PrestaShop\PrestaShop\Adapter\Product\ProductImagePathFactory;
+use PrestaShop\PrestaShop\Core\Configuration\UploadSizeConfigurationInterface;
+use PrestaShop\PrestaShop\Core\Hook\HookDispatcherInterface;
+use PrestaShop\PrestaShop\Core\Image\Exception\CannotUnlinkImageException;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\ImageUploadException;
+
+/**
+ * Uploads product image to filesystem
+ */
+class ProductImageUploader extends AbstractImageUploader
+{
+    /**
+     * @var UploadSizeConfigurationInterface
+     */
+    private $uploadSizeConfiguration;
+
+    /**
+     * @var ProductImagePathFactory
+     */
+    private $productImagePathFactory;
+
+    /**
+     * @var int
+     */
+    private $contextShopId;
+
+    /**
+     * @var ImageGenerator
+     */
+    private $imageGenerator;
+
+    /**
+     * @var HookDispatcherInterface
+     */
+    private $hookDispatcher;
+
+    /**
+     * @var bool
+     */
+    private $isLegacyImageMode;
+
+    /**
+     * @param UploadSizeConfigurationInterface $uploadSizeConfiguration
+     * @param ProductImagePathFactory $productImagePathFactory
+     * @param int $contextShopId
+     * @param ImageGenerator $imageGenerator
+     * @param HookDispatcherInterface $hookDispatcher
+     * @param bool $isLegacyImageMode
+     */
+    public function __construct(
+        UploadSizeConfigurationInterface $uploadSizeConfiguration,
+        ProductImagePathFactory $productImagePathFactory,
+        int $contextShopId,
+        ImageGenerator $imageGenerator,
+        HookDispatcherInterface $hookDispatcher,
+        bool $isLegacyImageMode
+    ) {
+        $this->uploadSizeConfiguration = $uploadSizeConfiguration;
+        $this->productImagePathFactory = $productImagePathFactory;
+        $this->contextShopId = $contextShopId;
+        $this->imageGenerator = $imageGenerator;
+        $this->hookDispatcher = $hookDispatcher;
+        $this->isLegacyImageMode = $isLegacyImageMode;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function upload(Image $image, string $filePath): void
+    {
+        $this->createDestinationDirectory($image);
+        $destinationPath = $this->productImagePathFactory->getBasePath($image);
+        $this->uploadFromTemp($filePath, $destinationPath);
```

Comment:
TODO: still don't like this. As mentioned it will remove the source image, but in handler scope we are not ensuring that it is really image from tmp folder.
If someone uses this handler and passes source file straight from some system dir it will try to delete file from someone's machine :/

Reply:
should I simply copy image and leave tempnam do its job (be collected by gc) :thinking:

Reply:
> If someone uses this handler and passes source file straight from some system dir it will try to delete file from someone's machine :/

@zuk3975 isn't it the same for any function that moves a file ? If you use `rename()` php native function targeting a protected file you'll move it. It's a function with huge power.

It's also the sysadmin responsibility to ensure critical files are protected using the right access rules.

We should not "protect user against its own stupidity". Just like Bash does not prevent you from running `rm -rf /`.

What we must do however is to try to protect user against mistyped inputs. For example a user thinks PrestaShop expects a relative path while we need an absolute path. Or user simply provides a `$filepath` to a file that does not exist.

We must check if input is _valid_ not if it makes sense 😄  .